sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala

```diff
@@ -501,4 +513,88 @@ trait Row extends Serializable {
   private def getAnyValAs[T <: AnyVal](i: Int): T =
     if (isNullAt(i)) throw new NullPointerException(s"Value at index $i is null")
     else getAs[T](i)
+
+  /** The compact JSON representation of this row. */
+  def json: String = compact(jsonValue)
```

Comment:
@hvanhovell, how about reusing `JacksonGenerator` in our JSON datasource?

Reply:
There's `pretty` option for `prettyJson` too.

Reply:
Ah, right, schema can be unknown ..

Reply:
Well you still need the schema. The main reason for not using Jackson generator is that we need to convert back to an internal row and this is super slow.

Reply:
Hm, this API looks already pretty slow though, and I suspect this API should be called in a critical path .. ?
If it's supposed to be used in a critical path, we might rather have to provide a API to make a convert function given schema (so that we avoid type dispatch for every row).

One rather minor concern is that the JSON representation for a row seems different comparing to JSON datasource. e.g.)  https://github.com/apache/spark/pull/26013/files#r331463832 and https://github.com/apache/spark/pull/26013/files#diff-78ce4e47d137bbb0d4350ad732b48d5bR576-R578

and here a bit duplicates the codes ..

Reply:
So two things to consider here.

I want to use this in StreamingQueryProgress right? All the JSON serialization there is based on the json4s AST and not strings (which is what JacksonGenerator produces). 

There is a difference between it being slow, and what you are suggesting. The latter being crazy inefficient. Let's break that down:
- Row to InternalRow conversion. You will need to create a converter per row because there is currently no way we can safely cache a converter. You can either use `ScalaReflection` or `RowEncoder` here, the latter is particularly bad because it uses code generation (which takes in the order of mills and which is weakly cached on the driver).
- Setting up the JacksonGenerator, again this is uncached and we need to set up the same thing for each tuple. 
- Generating the string.
Do you see my point here? Or shall I write a benchmark?

Reply:
There's one API case we dropped performance improvement in `Row` as an example.

```scala
  @deprecated("This method is deprecated and will be removed in future versions.", "3.0.0")
  def merge(rows: Row*): Row = {
    // TODO: Improve the performance of this if used in performance critical part.
    new GenericRow(rows.flatMap(_.toSeq).toArray)
  }
```

Do you mind if I ask to add `@Unstable` or `@Private` for these new APIs instead just for future improvement in case, with `@since` tag?

Reply:
I will mark them as @Unstable. @Private is debatable, because it is not really meant as an internal only API.